common/tools/dev-tool/src/framework/check.ts

```diff
@@ -0,0 +1,297 @@
+  }
+}
+
+export function isCheckFailedError(e: unknown): e is CheckFailedError {
+  return (e as any)?.name === "CheckFailedError";
+}
+
+/**
+ * Check that condition is true.
+ * @param condition - condition to check
+ * @param message - short message to display
+ * @param detail - more details to display about the failure; can be multiple lines
+ */
+export function check(condition: boolean, message: string, detail?: string): void {
+  if (!condition) {
+    throw new CheckFailedError(message, detail);
+  }
+}
+
+export interface ScriptCheckOptions extends Omit<CheckOptions, "hasFix"> {
+  /**
+   * Command to be ran in order to run the check.
+   *
+   * If the command exits with a zero status code, that indicates the check is successful.
+   * If the command exits with a non-zero status code, the output of the command will be reported
+   * to provide details about the failure.
+   */
+  checkCommand: string;
+
+  /**
+   * Command to be ran to attempt to fix the check.
+   * If undefined, this check will not be fixable.
+   */
+  fixCommand?: string;
+}
+
+/**
+ * Create a check that runs a CLI command.
+ *
+ * @param options - options for creating the check, including the command to run, and, optionally, a command which can be run to attempt to fix the failing check.
+ * @returns - the created check
+ */
+export function scriptCheck(options: ScriptCheckOptions): Check {
+  return {
+    ...options,
+    hasFix: Boolean(options.fixCommand),
+    async check({ fix, project, verbose }) {
+      if (fix) {
+        check(options.fixCommand !== undefined, "fix command must be defined");
+        const { exitCode, output } = await run(options.fixCommand!, {
+          cwd: project.path,
+          captureExitCode: true,
+          stdio: verbose ? "inherit" : undefined,
+          captureOutput: true,
+        });
+        check(exitCode === 0, `Check output exit code ${exitCode}`, output);
+      } else {
+        const { exitCode, output } = await run(options.checkCommand, {
+          cwd: project.path,
+          captureExitCode: true,
+          stdio: verbose ? "inherit" : undefined,
+          captureOutput: true,
+        });
+        check(exitCode === 0, `Check output exit code ${exitCode}`, output);
+      }
+    },
+  };
+}
+
+/**
+ * Context passed to the check() function of a package.json check.
+ */
+export interface PackageJsonCheckContext extends CheckContext {
+  /**
+   * JavaScript object representation of the project's package.json file.
+   * This object may be mutated to cause check failures if the mutation results in a change, or to fix the check if in fix mode.
+   */
+  packageJson: PackageJson;
+}
+
+export interface PackageJsonCheckOptions extends CheckOptions {
+  /**
+   * Run the check. The package.json is provided as a JavaScript object at the top level of the context, `packageJson`.
+   * This object may be mutated to cause check failures if the mutation results in a change, or to fix the check if in fix mode.
```

Comment:
I figured out how this worked by considering the examples, but it's maybe a little subtle that it relies on mutation in order to reuse fix logic for checking. Perhaps another way around this would be to allow defining *only* a fix function for PackageJsonChecks and then engaging this behavior?

Reply:
Thanks for the feedback. I changed a couple of things; I think it's a lot clearer but let me know what you think:

- package.json checks can now define a `fix` function, a `check` function, or both
  - if only specifying a fix function, when running the check in check mode then the check fails if the resulting package.json is meaningfully different (so similar behavior as now)
  - if only specifying a check function, the check works as normal and no fix is available
  - if both are specified, the check function is run in check mode, and the fix function is run in fix mode
- the `fix` function returns the updated `PackageJson` object instead of mutating the input `packageJson` in place. I think that's quite a bit clearer than mutating the input.